### PR TITLE
[*.c] Resolve some unused value in initial assignments; Fix some types…

### PR DESCRIPTION
**Pull request by SamuelMarks**
Found via my linter. There are plenty more where this comes from. Not sure if you want a larger number of changes to be added to this one PR so you can squash it into one commit? - Or maybe these changes impact too many files with too little benefit so just close the PR?

### CODE REVIEW

**Comment by bagder** [comment]
# Ubuntu
~~~
altsvc.c: In function ‘altsvc_add’:
altsvc.c:158:15: error: format ‘%u’ expects argument of type ‘unsigned int *’, but argument 11 has type ‘int *’ [-Werror=format=]
  158 |               "%" MAX_ALTSVC_ALPNLENSTR "s %" MAX_ALTSVC_HOSTLENSTR "s %u "
      |               ^~~
......
  163 |               date, &persist, &prio);
      |                               ~~~~~
      |                               |
      |                               int *
altsvc.c:160:56: note: format string is defined here
  160 |               "\"%" MAX_ALTSVC_DATELENSTR "[^\"]\" %u %u",
      |                                                       ~^
      |                                                        |
      |                                                        unsigned int *
      |                                                       %u
cc1: all warnings being treated as errors
~~~
# Windows
~~~
2>.\telnet.c(1183) : error C2220: warning treated as error - no 'object' file generated
2>.\telnet.c(1183) : warning C4310: cast truncates constant value
2>.\telnet.c(1198) : warning C4310: cast truncates constant value
2>.\telnet.c(1199) : warning C4310: cast truncates constant value

...

2>.\conncache.c(410) : error C2220: warning treated as error - no 'object' file generated
2>.\conncache.c(410) : warning C4018: '>' : signed/unsigned mismatch

File: lib/mqtt.c [hunk 1]
```diff
@@ -441,7 +441,7 @@ static CURLcode mqtt_read_publish(struct Curl_easy *data, bool *done)
   size_t remlen;
   struct mqtt_conn *mqtt = &conn->proto.mqtt;
   struct MQTT *mq = data->req.p.mqtt;
-  unsigned char packet;
+  int packet;
```
Comment (bagder) [inline lib/mqtt.c:444]: Why this change? It only stores a byte/char in there?

File: lib/telnet.c [hunk 2]
```diff
@@ -1172,20 +1172,20 @@ static CURLcode send_telnet_data(struct Curl_easy *data,
                                  char *buffer, ssize_t nread)
 {
   ssize_t escapes, i, outlen;
-  unsigned char *outbuf = NULL;
+  char *outbuf = NULL;
```
Comment (bagder) [inline lib/telnet.c:1175]: This change *adds* to the total number of typecasts, so why change it?

File: src/tool_getparam.c [hunk 3]
```diff
@@ -1370,13 +1370,13 @@ ParameterError getparameter(const char *flag, /* f or -long-flag */
          */
         const char *p = strchr(nextarg, '=');
         size_t nlen;
-        char is_file;
+        size_t is_file;
```
Comment (bagder) [inline src/tool_getparam.c:1373]: Why? It reads a single character from a buffer into the variable

File: tests/server/socksd.c [hunk 4]
```diff
@@ -169,7 +169,7 @@ static unsigned char byteval(char *value)
 static unsigned short shortval(char *value)
 {
   unsigned long num = strtoul(value, NULL, 10);
-  return num & 0xffff;
+  return (short)num & 0xffff;
```
Comment (bagder) [inline tests/server/socksd.c:172]: I think this is curious. We build on *many* compilers and none of them seems to have a problem with this. On what basis do we need a typecast there?

Comment (bagder) [reply tests/server/socksd.c:172]: And if we need one, shouldn't be it `unsigned short` ?

File: lib/mqtt.c (hunk 1 above)
Comment (SamuelMarks) [reply lib/mqtt.c:444]: Again, this was just stuff my linter picked up and I changed the type pretty blindly to make the linter play nice.

Seeing as you're actually open to accepted this PR, I'm putting some more time and effort in now to get it up to scratch, and ensure all the tests pass, and that the total number of warnings is reduced.

File: lib/telnet.c (hunk 2 above)
Comment (SamuelMarks) [reply lib/telnet.c:1175]: The main issue here is that `send_telnet_data` accepts a `char *buffer` not an `unsigned char *buffer`. So one way to remove all the casts is to change the interface. Would you accept such a change?

File: src/tool_getparam.c (hunk 3 above)
Comment (SamuelMarks) [reply src/tool_getparam.c:1373]: Agreed, I'll change it back. What made me go this direction was this line `nlen = is_file = 0;` made my linter complain, because obviously `0` isn't a `char` (but `'\0'` is). Regardless this is a common C-ism, so I'll revert it.

File: tests/server/socksd.c (hunk 4 above)
Comment (SamuelMarks) [reply tests/server/socksd.c:172]: Agreed, it should've been `unsigned short`. Again, we're implicitly truncating a `long`, so the linter complained. With the cast at least its explicit (yeah yeah, I know it's just a two line func)

**Comment by jay** [comment]
It's a common pattern to initialize result to CURLE_OK, or other variables to 0, even if not evaluated before being reassigned. Also IMO `if(curl_global_init(CURL_GLOBAL_ALL) != CURLE_OK)` is clearer than `if(curl_global_init(CURL_GLOBAL_ALL) != res)`

**Comment by jzakrzewski** [comment]
I'm a big fan of initialising at the point of declaration whenever possible. Even if the value isn't actually used afterwards.

**Comment by SamuelMarks** [comment]
@jay Some functions can never return `CURLE_OK`, so it's misleading to have the preluding boilerplate. Although it's a micro optimisation—if ever there was one—leaving variables unassigned is ostensibly more efficient. Also it's just basic housekeeping, in my code I would litter `const` everywhere possible. I would also remove all allocations which can be determined earlier on, e.g., with the exception of unicode filesystems one can infer the maximum path name and "allocate" that on the stack, in a `static` or just `const` character array.

So this PR is just the first of many "cleanup" contributions, which should hopefully improve the overall project (over time, naturally this first contribution will have a negligible impact).

File: lib/telnet.c (hunk 2 above)
Comment (bagder) [reply lib/telnet.c]: If it cleans up things, sure!

**Comment by bagder** [comment]
> I'm a big fan of initialising at the point of declaration whenever possible. Even if the value isn't actually used afterwards.

Experience also tells us that this habit makes it less likely that we introduce a bug when we in the future insert or remove lines that may or may not alter the sequence of assignments. So I too don't mind a few superfluous initializations.

File: tests/server/socksd.c [hunk 5]
```diff
@@ -169,7 +169,7 @@ static unsigned char byteval(char *value)
 static unsigned short shortval(char *value)
 {
   unsigned long num = strtoul(value, NULL, 10);
-  return (short)num & 0xffff;
+  return (unsigned short)num & 0xffff;
```
Comment (bagder) [inline tests/server/socksd.c:172]: If no compiler complains on this **and** it actually ANDs the value with `0xffff ` isn't this rather your linter that is overly picky?

Comment (SamuelMarks) [reply tests/server/socksd.c:172]: Well, the function is converting a string representation of a number into a `long`, and then is potentially truncating it (yes yes I know what the `&` does). So even though compilers may not complain about this, it is still implicit, and it's better to be explicit.

**Comment by SamuelMarks** [comment]
@bagder This codebase is pretty large. With that in mind, enabling as many "strict-mode" style compiler parameters, enforcing C89, and adding many tools to the CI layer is a good way to reduce bugs, decrease chances of security vulnerabilities; encouraging overall better quality contributions and maintenance.

So you could keep adding tools that trace allocations, model program flow, &etc &etc. That's a good idea in fact. Sure, you could have an exception list, like ignoring: unused values; unused parameters; type mismatching; and undefined behaviour. But are you sure you want that?

Wouldn't it be better if the tooling pointed out all these potential issues, and when one errs—[like you referenced] and returns a variable with an uninitalised value—that the tool will complain?

File: tests/server/socksd.c (hunk 5 above)
Comment (bagder) [reply tests/server/socksd.c:172]: >  it is still implicit, and it's better to be explicit.

Only to a certain extent.

It is also good to not write more code than we need to and to make the code easier on the eye for the reader. And typecasts do make a rather big dent in the flow when you read code I think. In my mind, code should have as few typecasts as possible.

Also: the function just above `shortval` in the same file is called `byteval` and it has the exact same logic but for an `unsigned char`. What made your linter be happy with that?

**Comment by bagder** [comment]
> adding many tools to the CI layer is a good way to reduce bugs, decrease chances of security vulnerabilities; encouraging overall better quality contributions and maintenance

Yes, we add and run *many* tools on the code to help us detect mistakes. But...

We only add and use such tools that do the right thing. There are (several) tools we don't use in automation simply because they're overly "whiny" or complains on things we don't agree with. For example assigning a value to a variable in the declaration of it, even if it gets assigned again before actually used. That's not an error for us.

**Comment by bagder** [comment]
This seems to have stalled while containing a lot of changes we don't want to merge. Closing. Feel free to reopen or submit a new PR if it awakes.
